packages/zero-cache/src/services/statz.ts

```diff
@@ -0,0 +1,359 @@
+        percentile_cont(0.99) WITHIN GROUP (ORDER BY row_count) AS "p99",
+        MIN(row_count) AS "min",
+        MAX(row_count) AS "max",
+        AVG(row_count) AS "avg"
+      FROM client_group_counts;`,
+      ],
+      [
+        // check for AST blowup due to DNF conversion.
+        'ast sizes',
+        sql`SELECT
+        percentile_cont(0.25) WITHIN GROUP (ORDER BY length("clientAST"::text)) AS "25th_percentile",
+        percentile_cont(0.5) WITHIN GROUP (ORDER BY length("clientAST"::text)) AS "50th_percentile",
+        percentile_cont(0.75) WITHIN GROUP (ORDER BY length("clientAST"::text)) AS "75th_percentile",
+        percentile_cont(0.9) WITHIN GROUP (ORDER BY length("clientAST"::text)) AS "90th_percentile",
+        percentile_cont(0.95) WITHIN GROUP (ORDER BY length("clientAST"::text)) AS "95th_percentile",
+        percentile_cont(0.99) WITHIN GROUP (ORDER BY length("clientAST"::text)) AS "99th_percentile",
+        MIN(length("clientAST"::text)) AS "minimum_length",
+        MAX(length("clientAST"::text)) AS "maximum_length",
+        AVG(length("clientAST"::text))::integer AS "average_length",
+        COUNT(*) AS "total_records"
+      FROM ${sql(schema)}."queries";`,
+      ],
+      [
+        // output the hash of the largest AST
+        'biggest ast hash',
+        sql`SELECT "queryHash", length("clientAST"::text) AS "ast_length"
+      FROM ${sql(schema)}."queries"
+      ORDER BY length("clientAST"::text) DESC
+      LIMIT 1;`,
+      ],
+      [
+        'total active queries per client and client group',
+        numQueriesPerClientGroup(true),
+      ],
+      [
+        'total inactive queries per client and client group',
+        numQueriesPerClientGroup(false),
+      ],
+      [
+        'total rows per client group',
+        sql`SELECT "clientGroupID", COUNT(*) as "c" FROM ${sql(
+          schema,
+        )}."rows" GROUP BY "clientGroupID" ORDER BY "c" DESC`,
+      ],
+      [
+        'num rows per query',
+        sql`SELECT
+        k.key AS "queryHash",
+        COUNT(*) AS row_count
+      FROM ${sql(schema)}."rows" r,
+      LATERAL jsonb_each(r."refCounts") k
+      GROUP BY k.key
+      ORDER BY row_count DESC;`,
+      ],
+    ] satisfies [
+      name: string,
+      query: ReturnType<ReturnType<typeof pgClient>>,
+    ][],
+    out,
+  );
+
+  out.write('ZQL (without permissions) for each query:');
+  const queryAsts =
+    await sql`SELECT "queryHash", "clientAST" FROM ${sql(schema)}."queries"`;
+
+  const seenQueries = new Set<string>();
+  const parseFailures: string[] = [];
+  for (const row of queryAsts) {
+    const {queryHash, clientAST} = row;
+    if (seenQueries.has(queryHash)) {
+      continue;
+    }
+    seenQueries.add(queryHash);
+
+    try {
+      const zql = clientAST.table + astToZQL(clientAST);
+      out.write('HASH:' + queryHash);
+      out.write('ZQL:' + zql + '\n');
+    } catch (e) {
+      parseFailures.push(queryHash);
+    }
```

Comment:
Not sure we should dump these. It could be quite a lot of data.

Reply:
Yeah :-/. Let's see how it goes. Seems useful.